### ISSUE TITLE
provide an option to skip initializing large objects in GC

### ISSUE THREAD

**Issue by Maoni0** [dotnet/coreclr#17324]
sometimes it may be desirable to not have GC to clear memory when it hands an object out. I am using this issue to track providing such an option for large objects in GC which is pretty simple - if GCHeap::Alloc is called with this intention indicated in the flag GC will not clear the memory. there's also a part in the higher level component that will need to expose such a flag and pass it down to the alloc helper which will be tracked in a separate issue.

**Comment by ygc369** [comment]
@Maoni0
Why does GC have to clear memory? I think that GC needn't clear memory for both large and small objects.
Memory shouldn't be cleared until it is reallocated, should it?
So not clearing memory should be default behavior for GC, then why there should be an option for it?

**Comment by colgreen** [comment]
Thought: Could/should this functionality be within a statement block marked as unsafe, as per pinned references and such?

**Comment by adamsitnik** [comment]
> Why does GC have to clear memory? 

@ygc369 when you allocate an array of value types .NET Runtime guarantees that all the elements inside it are going to have the default value. It does so by zeroing the memory.

This is why following test is always passing:

```cs
int[] array = new int[100]; // allocates memory and zeroes it
Xunit.Assert.All(array, element => Assert.Equal(0, element));
```

but sometimes you don't need it, for example, you want to initialize every element with its own index and having some garbage inside it before the first write is fine

```cs
int[] array = new int[100]; // allocates memory and zeroes it 
for(int i = 0; i < array.Length; i++)
  array[i] = i;
```

**Comment by VSadov** [comment]
I think this issue is addressed in #24096,  there is an option to not clear now.

I will keep https://github.com/dotnet/corefx/issues/31787 open to follow up with making the API public.
